### PR TITLE
Feature/add services

### PR DESCRIPTION
**Pull request by DmitriyG0r4anin**

### CODE REVIEW

**Review by MaliyOna — APPROVED** [review]

File: Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        var cache = await distributedCache.GetDataFromCacheAsync<VehicleClientHistoryModel>(key, cancellationToken);
+
+        if (cache is not null)
+            return cache;
+
+        var vch = await repository.GetByIdAsync(id, cancellationToken);
+
+        var vchModel = vch.Adapt<VehicleClientHistoryModel>();
+
+        var cacheLifetime = TimeSpan.FromMinutes(10);
```
Comment (5Ka-me) [inline Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs]: Move lifetime to constant or env.variable

File: Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        var newVchModel = vch.Adapt<VehicleClientHistoryModel>();
+
+        return newVchModel;
+    }
+
+    public async Task<VehicleClientHistoryModel> UpdateAsync(VehicleClientHistoryModel vchModel, CancellationToken cancellationToken)
+    {
+        var newVchModel = await repository.GetByIdAsync(vchModel.Id, cancellationToken)
+            ?? throw new NotFoundException(ExceptionMessages.NotFound(nameof(VehicleClientHistoryEntity), vchModel.Id));
```
Comment (5Ka-me) [inline Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs]: Won't it throw an error in repo by EF?

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        return vchModelToReturn;
+    }
+
+    public async Task DeleteAsync(Guid id, CancellationToken cancellationToken)
+    {
+        var modelName = await repository.GetByIdAsync(id, cancellationToken)
+            ?? throw new NotFoundException(ExceptionMessages.NotFound(nameof(VehicleClientHistoryEntity), id));
+
+        await repository.RemoveAsync(modelName, cancellationToken);
```
Comment (5Ka-me) [inline Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs]: Wouldn't it be better to move this logic to repo?

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs (hunk 3 above)
Comment (DmitriyG0r4anin) [reply Rent.BusinessLogic/Services/Implementations/VehicleClientHistoryService.cs]: will be resolved in specified branch

File: Rent.BusinessLogic/Extensions/Cache.cs [hunk 4]
```diff
@@ -17,17 +19,19 @@ internal static class Cache
...
     }
 
-    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, TimeSpan cacheLifetime, string key, CancellationToken cancellationToken)
+    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, string key, CancellationToken cancellationToken)
     {
         if (data is null)
             return;
 
+        if (!int.TryParse(lifetime, out var lifetimeMinutes))
+            return;
```
Comment (5Ka-me) [inline Rent.BusinessLogic/Extensions/Cache.cs]: Don't you think it is better to throw an exception instead of returning null?

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: Rent.BusinessLogic/Extensions/Cache.cs [hunk 5]
```diff
@@ -17,17 +19,19 @@ internal static class Cache
...
 
-    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, TimeSpan cacheLifetime, string key, CancellationToken cancellationToken)
+    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, string key, CancellationToken cancellationToken)
     {
-        if (data is null)
+        if (data is null || lifetime is null)
             return;
 
+        if (!double.TryParse(lifetime, out var lifetimeMinutes))
+            throw new InvalidDataException("Cache lifetime is not valid");
```
Comment (5Ka-me) [inline Rent.BusinessLogic/Extensions/Cache.cs]: The same. Check another repo

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

**Review by 5Ka-me — APPROVED** [review]

**Pull request merged**